### PR TITLE
Adds VideoEncoder

### PR DESCRIPTION
**Pull request by rachelandrew**
Here is VideoEncoder.

Reviewer: @jpmedley

### CODE REVIEW

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  - API
+  - Method
+  - Reference
+  - configure
+  - VideoEncoder
+browser-compat: api.VideoEncoder.configure
+---
+{{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
+
+The **`configure`** method of the {{domxref("VideoEncoder")}} interface enqueues a control message to configure the audio decoder for decoding chunks.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:14]:
```suggestion
The **`configure()`** method of the {{domxref("VideoEncoder")}} interface enqueues a control message to configure the audio decoder for decoding chunks.
```

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    - `width`{{Optional_Inline}}
+      - : An integer representing the width of each output {{domxref("EncodedVideoChunk")}} in pixels, before any ratio adjustments.
+    - `height`{{Optional_Inline}}
+      - : An integer representing the height of each output {{domxref("EncodedVideoChunk")}} in pixels, before any ratio adjustments.
+    - `displayWidth`{{Optional_Inline}}
+      - : An integer representing the intended display width of each output {{domxref("EncodedVideoChunk")}} in pixels when displayed.
+    - `displayHeight`{{Optional_Inline}}
+      - : An integer representing the vertical dimension of each output {{domxref("EncodedVideoChunk")}} in pixels when displayed.
+    - `hardwareAcceleration`
+      - : An hint that configures the hardware accelaration method of this codec. One of:
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:37]:
```suggestion
      - : A hint that configures the hardware acceleration method of this codec. One of:
```

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        - `"no-preference"`
+        - `"prefer-hardware"`
+        - `"prefer-software"`
+    - `bitrate`
+      - : An integer containing the average bitrate of the encoded video in units of bits per second.
+    - `framerate`
+      - : An integer containing the expected framerate in frames per second.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:44]:
```suggestion
      - : An integer containing the expected frame rate in frames per second.
```

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    - `alpha`
+      - : A string indicating whether the alpha component of the VideoFrame inputs should be kept or discarded prior to encoding, defaults to `"discard"`. One of:
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:46]:
```suggestion
      - : A string indicating whether the alpha component of the `VideoFrame` inputs should be kept or discarded prior to encoding. One of:
```

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        - `"discard"`
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:47]:
```suggestion
        - `"discard"` (default)
```

File: files/en-us/web/api/videoencoder/configure/index.md (hunk 4 above)
Comment (jpmedley) [reply files/en-us/web/api/videoencoder/configure/index.md:46]: This goes with the next one. This optimizes for scanning.

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        - `"keep"`
+    - `scalabilityMode`
+      - : A {{domxref("DOMString", "string")}} containing an encoding scalability mode identifier as defined in [WebRTC](https://w3c.github.io/webrtc-svc/#scalabilitymodes*).
+    - `bitrateMode`
+      - : A string containing a bitrate mode, defaulting to `"variable"`. One of:
+        - `"constant"`
+        - `"variable"`
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:54]:
```suggestion
        - `"variable"` (default)
```

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    - `framerate`
+      - : An integer containing the expected framerate in frames per second.
+    - `alpha`
+      - : A string indicating whether the alpha component of the VideoFrame inputs should be kept or discarded prior to encoding, defaults to `"discard"`. One of:
+        - `"discard"`
+        - `"keep"`
+    - `scalabilityMode`
+      - : A {{domxref("DOMString", "string")}} containing an encoding scalability mode identifier as defined in [WebRTC](https://w3c.github.io/webrtc-svc/#scalabilitymodes*).
+    - `bitrateMode`
+      - : A string containing a bitrate mode, defaulting to `"variable"`. One of:
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:52]:
```suggestion
      - : A string containing a bitrate mode. One of:
```

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        - `"constant"`
+        - `"variable"`
+    - `latencyMode`
+      - : A string containing a value that configures the latency behavior of this codec, defaulting to `"quality"`. One of:
+        - `"quality"`
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:57]:
```suggestion
        - `"quality"` (default)
```

File: files/en-us/web/api/videoencoder/configure/index.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        - `"discard"`
+        - `"keep"`
+    - `scalabilityMode`
+      - : A {{domxref("DOMString", "string")}} containing an encoding scalability mode identifier as defined in [WebRTC](https://w3c.github.io/webrtc-svc/#scalabilitymodes*).
+    - `bitrateMode`
+      - : A string containing a bitrate mode, defaulting to `"variable"`. One of:
+        - `"constant"`
+        - `"variable"`
+    - `latencyMode`
+      - : A string containing a value that configures the latency behavior of this codec, defaulting to `"quality"`. One of:
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/configure/index.md:56]:
```suggestion
      - : A string containing a value that configures the latency behavior of this codec. One of:
```

File: files/en-us/web/api/videoencoder/encode/index.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+VideoEncoder.encode(frame);
+VideoEncoder.encode(frame, options);
+```
+
+### Parameters
+
+- `frame`
+  - : A {{domxref("VideoFrame")}} object.
+- `options`{{Optional_Inline}}
+  - : A dictionary object containing the following member:
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/encode/index.md:28]:
```suggestion
  - : An object containing the following member:
```

File: files/en-us/web/api/videoencoder/encode/index.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+### Return Value
+
+{{jsxref("Undefined")}}.
+
+### Exceptions
+
+- {{domxref("DOMException")}} `InvalidStateError`
+  - : Thrown if the {{domxref("VideoEncoder.state","state")}} is not `"configured"`.
+- {{domxref("DOMException")}} `DataError`
+  - : Thrown if the `chunk` is unable to be decoded due to relying on other frames for decoding.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/encode/index.md:41]:
```suggestion
  - : Thrown if the `chunk` cannot be decoded due to relying on other frames for decoding.
```

File: files/en-us/web/api/videoencoder/index.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+title: VideoEncoder
+slug: Web/API/VideoEncoder
+tags:
+  - API
+  - Interface
+  - Reference
+  - VideoEncoder
+browser-compat: api.VideoEncoder
+---
+The **`VideoEncoder`** interface of the {{domxref('WebCodecs API','','',' ')}} decodes {{domxref("VideoFrame")}} objects.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/index.md:11]:
```suggestion
The **`VideoEncoder`** interface of the {{domxref('WebCodecs API','','','true')}} decodes {{domxref("VideoFrame")}} objects.
```

File: files/en-us/web/api/videoencoder/state/index.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  - API
+  - Property
+  - Reference
+  - state
+  - VideoEncoder
+browser-compat: api.VideoEncoder.state
+---
+{{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
+
+The **`state`**  property of the {{domxref("VideoEncoder")}} interface returns the current state of the underlying codec.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/state/index.md:14]:
```suggestion
The **`state`** property of the {{domxref("VideoEncoder")}} interface returns the current state of the underlying codec.
```

File: files/en-us/web/api/videoencoder/videoencoder/index.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+## Syntax
+
+```js
+new VideoEncoder(init);
+```
+
+### Parameters
+
+- `init`
+  - : A dictionary object containing two required callbacks.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/videoencoder/index.md:24]:
```suggestion
  - : An object containing two required callbacks.
```

File: files/en-us/web/api/videoencoder/videoencoder/index.md [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    - `output`
+      - : A callback which is passed a single argument of an {{domxref("VideoFrame")}} object.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/videoencoder/index.md:26]:
```suggestion
      - : A callback which takes a {{domxref("VideoFrame")}} object as its only argument.
```

File: files/en-us/web/api/videoencoder/videoencoder/index.md [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    - `error`
+      - : A callback which is passed a single argument of the error that has been thrown.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/videoencoder/index.md:28]:
```suggestion
      - : A callback which takes an {{jsxref("Error")}} object as its only argument.
```

File: files/en-us/web/api/videoencoder/videoencoder/index.md [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+## Examples
+
+In the following example an `VideoEncoder` is created with the two required callback functions, one to deal with the decoded frame and the other to handle errors.
```
Comment (jpmedley) [inline files/en-us/web/api/videoencoder/videoencoder/index.md:32]:
```suggestion
In the following example a `VideoEncoder` is created with the two required callback functions, one to deal with the decoded frame and the other to handle errors.
```

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: files/en-us/web/api/videoencoder/state/index.md [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+  - API
+  - Property
+  - Reference
+  - state
+  - VideoEncoder
+browser-compat: api.VideoEncoder.state
+---
+{{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
+
+The **`state`** property of the {{domxref("VideoEncoder")}} interface returns the current state of the underlying codec.
```
Comment (rachelandrew) [inline files/en-us/web/api/videoencoder/state/index.md:14]:
```suggestion
The **`state`** read-only property of the {{domxref("VideoEncoder")}} interface returns the current state of the underlying codec.
```

**Review by jpmedley — APPROVED** [review]

**Pull request merged**